include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h

```diff
@@ -646,24 +646,80 @@ __pattern_count(__hetero_tag<_BackendTag>, _ExecutionPolicy&& __exec, _Iterator
 // any_of
 //------------------------------------------------------------------------
 
+template <typename _Tuple, typename _UnaryTransformOp>
+struct __find_if_unary_transform_op
+{
+    _UnaryTransformOp __transform_op;
+
+    template <typename Arg>
+    _Tuple
+    operator()(const Arg& arg) const
+    {
+        return {__transform_op(std::get<0>(arg)), std::get<1>(arg)};
+    }
+};
+
+template <typename _Tuple, typename _IsFirst>
+struct __find_if_binary_reduce_op
+{
+    _Tuple
+    operator()(const _Tuple& op1, const _Tuple& op2) const
+    {
+        if (std::get<0>(op1) && std::get<0>(op2))
+        {
+            if constexpr (_IsFirst{})
+                return {true, std::min(std::get<1>(op1), std::get<1>(op2))};
+            else
+                return {true, std::max(std::get<1>(op1), std::get<1>(op2))};
+        }
+
+        return std::get<0>(op1) ? op1 : op2;
+    }
+};
+
 template <typename _BackendTag, typename _ExecutionPolicy, typename _Iterator, typename _Pred>
 bool
 __pattern_any_of(__hetero_tag<_BackendTag>, _ExecutionPolicy&& __exec, _Iterator __first, _Iterator __last,
                  _Pred __pred)
 {
-    if (__first == __last)
+    using _difference_type = typename ::std::iterator_traits<_Iterator>::difference_type;
+
+    const _difference_type __n = __last - __first;
+    if (__n == 0)
         return false;
 
-    using _Predicate = oneapi::dpl::unseq_backend::single_match_pred<_ExecutionPolicy, _Pred>;
+    using _result_type = oneapi::dpl::__internal::tuple<bool, _difference_type>;
+    const auto __init = _result_type{false, __n};
 
-    auto __keep = oneapi::dpl::__ranges::__get_sycl_range<__par_backend_hetero::access_mode::read, _Iterator>();
-    auto __buf = __keep(__first, __last);
+    // __counting_iterator_t - iterate position (index) in source data
+    using __counting_iterator_t = oneapi::dpl::counting_iterator<_difference_type>;
```

Comment:
I'm not following the reason to include the counting iterator and the rest of this that tracks the index of the output for `__pattern_any_of`, which only returns the `true` or `false` information, and discards the index information.  It seems we are adding lots of complexity and computation.
If we want to handle this with a transform_reduce, it seems like we can avoid the counting iterator, zipping, tuples, etc. 
 and just have a reduction operation `std::logical_and`.  

I should note that I'm not sure a reduction is what we want here necessarily.  There may be some better "early exit" routine which checks for `true` and short circuits remaining work if it encounters one.  As discussed though, this would need to balance contention on some synchronization mechanism with the potential savings.

Reply:
Good, point, fixed.